runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1434 @@
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "j9.h"
+#include "jclprots.h"
+#include "j9cp.h"
+#include "j9protos.h"
+#include "ut_j9jcl.h"
+#include "j9port.h"
+#include "jclglob.h"
+#include "jcl_internal.h"
+#include "util_api.h"
+#include "j9vmconstantpool.h"
+#include "ObjectAccessBarrierAPI.hpp"
+#include "objhelp.h"
+
+#include <string.h>
+#include <assert.h>
+
+#include "VMHelpers.hpp"
+
+extern "C" {
+
+#define MN_IS_METHOD		0x00010000
+#define MN_IS_CONSTRUCTOR	0x00020000
+#define MN_IS_FIELD			0x00040000
+#define MN_IS_TYPE			0x00080000
+#define MN_CALLER_SENSITIVE	0x00100000
+
+#define MN_REFERENCE_KIND_SHIFT	24
+#define MN_REFERENCE_KIND_MASK	0xF
+
+#define MN_SEARCH_SUPERCLASSES	0x00100000
+#define MN_SEARCH_INTERFACES	0x00200000
+
+void
+initImpl(J9VMThread *currentThread, j9object_t membernameObject, j9object_t refObject)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	J9Class* refClass = J9OBJECT_CLAZZ(currentThread, refObject);
+
+	jint flags = 0;
+	jlong vmindex = 0;
+	jlong target = 0;
+	j9object_t clazzObject = NULL;
+	j9object_t nameObject = NULL;
+	j9object_t typeObject = NULL;
+
+	if (refClass == J9VMJAVALANGREFLECTFIELD(vm)) {
+		J9JNIFieldID *fieldID = vm->reflectFunctions.idFromFieldObject(currentThread, NULL, refObject);
+		vmindex = (jlong)fieldID;
+		target = (jlong)fieldID->field;
+
+		flags = fieldID->field->modifiers & CFR_FIELD_ACCESS_MASK;
+		flags |= MN_IS_FIELD;
+		flags |= (J9_ARE_ANY_BITS_SET(flags, J9AccStatic) ? MH_REF_GETSTATIC : MH_REF_GETFIELD) << MN_REFERENCE_KIND_SHIFT;
+		J9UTF8 *name = J9ROMFIELDSHAPE_NAME(fieldID->field);
+		J9UTF8 *signature = J9ROMFIELDSHAPE_SIGNATURE(fieldID->field);
+
+		nameObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(name), (U_32)J9UTF8_LENGTH(name), J9_STR_INTERN);
+		typeObject = vm->memoryManagerFunctions->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(signature), (U_32)J9UTF8_LENGTH(signature), J9_STR_INTERN);
```

Comment:
`nameObject` is corrupted if a GC occurs during the second object allocation.  The object pointer needs to be pushed in a special frame or stored somewhere the GC can find it and refreshed after the second allocate

Reply:
fixed